### PR TITLE
BLD: replace uses of openblas_support with openblas wheels [wheel build]

### PR DESCRIPTION
**Pull request by mattip**
Use the openblas wheels instead of `tools/openblas_support.py` in
- wheel building
- nightly/stable openblas32 CI runs
- sdist

This will fix #25503 (unpin cibuildwheel)

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks Matti. This looks about right; the failures on 32-bit Windows, musllinux and macOS arm64 are all the same (per platform) - you may want to fix those on a single wheel build job on your own fork before running the whole battery of builds here.

**Comment by andyfaff** [comment]
Checkout https://github.com/numpy/numpy/wiki/Debugging-CI-guidelines for some suggestions on how to debug cirrus locally (won't help with macos arm 64 if you don't have your own Mac)

**Comment by mattip** [comment]
Mess. The scipy-openblas wheels differ from the tarballs. The tarballs we currently use have, in usr/local/lib, symlinks like
```
libscipy_openblas64_.a -> libscipy_openblas64_p-r0.3.24.dev.a
libscipy_openblas64_.so -> libscipy_openblas64_p-r0.3.24.dev.so
libscipy_openblas64_.so.0 -> libscipy_openblas64_p-r0.3.24.dev.so
libscipy_openblas64_p-r0.3.24.dev.so

```

But the real file `libscipy_openblas64_p-r0.3.24.dev.so` has an `SONAME` of `libscipy_openblas64_.so.0`, so after linking `_multiarray_umath.so` with the shared object it expects to find `libscipy_openblas64_.so.0` in the `LD_LIBRARY_PATH` (the names are changed but the idea is the same on macos). This makes dynamically linking the artifacts in the wheel useless.

I think we have to rework the scipy-openblas wheel build before we can use it here to 
- ship `libscipy_openblas64_.so` not `libscipy_openblas64_p-r0.3.24.dev.so`, and
- change the SONAME to `libscipy_openblas64_.so`

**Comment by rgommers** [comment]
That seems like the right change to make. There's no need for names like `p-r0.3.24` at all, and symlinks in wheels are bad.

**Review by matheussouza88 — APPROVED** [review]

File: tools/wheels/cibw_before_build.sh [hunk 1]
```diff
@@ -16,52 +15,57 @@ elif [[ $RUNNER_OS == "Windows" ]]; then
...
+if [[ $INSTALL_OPENBLAS ]] ; then
+    # PKG_CONFIG_PATH=$PROJECT_DIR/.openblas
+    mkdir -p $PKG_CONFIG_PATH/lib
+    python -m pip install scipy-openblas64
+    python -c "import scipy_openblas64; print(scipy_openblas64.get_pkg_config())" > $PKG_CONFIG_PATH/scipy-openblas.pc
+    # Explicitly link the openblas so instead of using the "import
+    # scipy-openblas64" preload trick
+    sed -e "s/Libs:.*$/Libs: \${extralib}/" -i.bak $PKG_CONFIG_PATH/scipy-openblas.pc
+    sed -e "s/-lm -lpthread -lgfortran -lquadmath//" -i.back $PKG_CONFIG_PATH/scipy-openblas.pc
+    rm $PKG_CONFIG_PATH/*.bak
```
Comment (mattip) [inline tools/wheels/cibw_before_build.sh:36]: Fixing this upstream in MacPython/openblas-libs#128
```suggestion
```

File: tools/wheels/cibw_before_build.sh [hunk 2]
```diff
@@ -16,52 +15,57 @@ elif [[ $RUNNER_OS == "Windows" ]]; then
...
+        # since the runtime loader will look for a file with this name
+        # objdump -p will output something like
+        # SONAME libscipy_openblas64_0.so
+        # on linux and
+        # name libscipy_openblas64_0.dylib (offset 24)
+        # on macos, so trim off the first and last parts of the line
+        soname=$(objdump -p $PKG_CONFIG_PATH/lib/libscipy* | grep -v "p-r" | grep libscipy | sed -e"s/^.*libsci/libsci/" | cut -f1 -d" ")
+        echo "linking shared object to '$soname'"
+        ln -s $PKG_CONFIG_PATH/lib/libscipy* $PKG_CONFIG_PATH/lib/$soname
     fi
```
Comment (mattip) [inline tools/wheels/cibw_before_build.sh:54]: Fixing this upstream in https://github.com/MacPython/openblas-libs/pull/128
```suggestion
```

**Comment by mattip** [comment]
Updated to use the newer 0.3.26.0.2 wheels with a single shared object

**Comment by rgommers** [comment]
Diff looks reasonable - just some wheel build jobs to make happy

**Comment by mattip** [comment]
@andyfaff do you know why the cirrus build sequence for wheels first builds NumPy, then calls cibuildwheel to build it again? It is failing since the cibw_before_build.sh does not overwrite `$PWD/.openblas` the second time it is called.

**Comment by mattip** [comment]
macos with python 3.11, 3.12 and pypy3.9 is failing but with python3.9 and 3.10 is passing. The problem seems to be that delocate is not seeing the `DYLD_LIBRARY_PATH` environment variable, so delocate does not find the scipy_openblas shared object. Did something change in the way delocate process the environment between the versions?

**Comment by andyfaff** [comment]
> do you know why the cirrus build sequence for wheels first builds NumPy, then calls cibuildwheel to build it again?

It shouldn't do. Can you link to a build log that shows this. When I look at e.g. the cron job on cirrus I can't see the two builds.
I did notice something wrong with the cirrus macos_arm64 build though, it shouldn't be installing gfortran from cibw_before_build, it should be using the mamba environment installed compiler for gfortran.

A quick reminder, there is a matrix of images that the wheel building occurs for. Once for all the pythons on sonoma, and once on monterey. The wheels on Sonoma should be using Accelerate for BLAS, those on Monterey are supposed to still use OpenBLAS.
So that's a second issue, the Sonoma builds for the last [cron job](https://cirrus-ci.com/task/6678350190608384?logs=build#L981) are finding and using scipy-openblas, when they should be using Accelerate. That needs to be fixed, the whole point of Accelerate is to reduce the size of wheels. It's important to fix this before the next release happens! (@charris, @rgommers). I don't actually know where scipy-openblas is being installed for that wheel build job.

It's not commented why delocate is pinned in the cibw_before_build script. I'll have a look into these issues in the next few days. It's going to be a slow ping time for me because I'm busy at work this week.

**Comment by mattip** [comment]
> Can you link to a build log that shows this. 

Here is [a build log](https://github.com/numpy/numpy/pull/25505/checks?check_run_id=21189476200) with 4 steps: clone, build, install_cibuildwheel, cibuildwheel.

> the Sonoma builds for the last [cron job](https://cirrus-ci.com/task/6678350190608384?logs=build#L981) are finding and using scipy-openblas

Those builds are probable not setting `INSTALL_OPENBLAS=false`, so it is being set by default here https://github.com/numpy/numpy/blob/d3345bb1f5682cc54850708dfed213b1ae530a6f/tools/wheels/cibw_before_build.sh#L20-L22

which then triggers the use of OpenBLAS

**Comment by andyfaff** [comment]
> clone, build, install_cibuildwheel, cibuildwheel

The naming of those build steps is a little cryptic. The first 'build' step is actually firing all the other steps off, not actually building numpy in it. It could be called something like 'initial setup' or something like that. I'm not sure why the cibw_before_build script is executed in that step, possibly cruft that can be cleaned up. Debugging all of this is probably better done locally, using the wiki instructions I wrote.

File: pyproject.toml [hunk 3]
```diff
@@ -155,22 +154,28 @@ manylinux-x86_64-image = "manylinux2014"
...
+LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/project/.openblas/lib"
 
 [tool.cibuildwheel.macos]
 # universal2 wheels are not supported (see gh-21233), use `delocate-fuse` if you need them
 archs = "x86_64 arm64"
 test-skip = "*_universal2:arm64"
-environment = {RUNNER_OS="macOS"}
+[tool.cibuildwheel.macos.environment]
+# In cibuildwheel, {project} is $PWD
+RUNNER_OS="macOS"
```
Comment (andyfaff) [inline pyproject.toml:168]: I think the variables set in this section are overwritten by https://github.com/numpy/numpy/blob/main/tools/ci/cirrus_wheels.yml#L94.  THey're set in the ci config script because we have to figure out what kind of system we're building on first.

**Comment by andyfaff** [comment]
Should be fixed by #25763

**Comment by mattip** [comment]
>  Debugging all of this is probably better done locally, using the wiki instructions I wrote.

Thanks I will try out [these instructions without a VM](https://github.com/numpy/numpy/wiki/Debugging-CI-guidelines#running-cirrus-ci-configurations-locally) on a mac machine. Or did you mean the containerized one?

**Comment by andyfaff** [comment]
If you have a macos m series I would try https://github.com/numpy/numpy/wiki/Debugging-CI-guidelines#local-build-of-macos-wheels-using-cibuildwheel-in-a-macos-tart-vm with the entries that youve seen fail.

**Comment by mattip** [comment]
Rebased on top of #25763 and added a fix for cibuildwheel not setting macos repair wheel environment variables as specified in the `pyproject.toml`

File: pyproject.toml (hunk 3 above)
Comment (mattip) [reply pyproject.toml:175]: I see. But even in a local build, with only the values fro the `pyproject.toml`, `DYLD_LIBRARY_PATH` is not percolating into the repair wheel command.

Comment (andyfaff) [reply pyproject.toml:175]: Does the way that scipy do it help? https://github.com/scipy/scipy/blob/a6587b6c1a806d384b292592423693e4c573882c/ci/cirrus_wheels.yml#L76

File: pyproject.toml [hunk 4]
```diff
@@ -155,22 +154,35 @@ manylinux-x86_64-image = "manylinux2014"
...
 [tool.cibuildwheel.macos]
 # universal2 wheels are not supported (see gh-21233), use `delocate-fuse` if you need them
 archs = "x86_64 arm64"
 test-skip = "*_universal2:arm64"
-environment = {RUNNER_OS="macOS"}
+# Not clear why the DYLD_LIBRARY_PATH is not passed through from the environment
+repair-wheel-command = [
+  "export DYLD_LIBRARY_PATH=$PWD/.openblas/lib",
+  "echo DYLD_LIBRARY_PATH $DYLD_LIBRARY_PATH",
+  "delocate-wheel --require-archs {delocate_archs} -w {dest_dir} -v {wheel}",
```
Comment (andyfaff) [inline pyproject.toml:170]: I wonder if we should specifically exclude the openblas lib from the delocate call for Sonoma wheels?

**Comment by mattip** [comment]
Hmm. The cirrus arm64 monterey builds are failing to find scipy-openblas, but the build is continuing to completion. Apparently there is some older blas somewhere in the image:
```
Run-time dependency scipy-openblas found: NO (tried pkgconfig)
Run-time dependency accelerate found: NO (tried system)
Found CMake: /opt/homebrew/bin/cmake (3.24.1)
WARNING: CMake Toolchain: Failed to determine CMake compilers state
Run-time dependency openblas found: NO (tried pkgconfig, pkgconfig, pkgconfig, system and cmake)
Run-time dependency flexiblas64 found: NO (tried pkgconfig, framework and cmake)
Run-time dependency blis found: NO (tried pkgconfig, framework and cmake)
Run-time dependency blas found: YES unknown
../numpy/meson.build:135: WARNING: Project targets '>=1.2.99' but uses feature introduced in '1.3.0': dep 'blas' custom lookup.
Message: BLAS symbol suffix: 
Run-time dependency accelerate found: NO (tried system)
Run-time dependency openblas found: NO (tried pkgconfig, pkgconfig, pkgconfig, system and cmake)
Run-time dependency flexiblas found: NO (tried pkgconfig, framework and cmake)
Run-time dependency lapack found: YES unknown
../numpy/meson.build:198: WARNING: Project targets '>=1.2.99' but uses feature introduced in '1.3.0': dep 'lapack' custom lookup.
``` 

On the other hand, the sonoma builds are using Accelerate (detected by looking in the logs).

More local debugging needed.

File: pyproject.toml (hunk 4 above)
Comment (mattip) [reply pyproject.toml:170]: It would be cleaner, yes. Setting `DYLD_LIBRARY_PATH` should be harmless, since 
- `$PWD/.openblas/lib` should not exist
- the build should use Accelerate, so delocate should not find any need for OpenBLAS.

Comment (andyfaff) [reply pyproject.toml:170]: I was thinking along the lines of it being a reinforcement for knowing the wheel is using Accelerate.

File: pyproject.toml (hunk 3 above)
Comment (andyfaff) [reply pyproject.toml:175]: People using cibw to build locally tend to be core developers figuring out how to keep infra running.

Comment (mattip) [reply pyproject.toml:175]: heh, yeah, makes sense. Who else would want wheels.

**Comment by mattip** [comment]
On `macosx_arm64 CIBW_BUILD:cp310-* cp311-* montery`, the meson build state is being cached between the two build runs (for cp310 and cp311). This is wrong, the build should start cleanly. I am seeing (cached) in the log, even though it has `--reconfigure` in the command line. This breaks the scipy-openblas build since the previous build uses the scipy-openblas files from python310 venv, and in the python 311 build that venv is gone:
<details><summary>detail from the log file</summary>

```
private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/build-env-9ygl4v_7/bin/python \
    /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/vendored-meson/meson/meson.py \
    setup --reconfigure /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build \
    /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/build \
    -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md -Dallow-noblas=false \
    --native-file=/private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/build/meson-python-native-file.ini
Cleaning... 182 files.
The Meson build system
Version: 1.2.99
Source dir: /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build
Build dir: /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/build
Build type: native build
Project name: NumPy
Project version: 2.0.0.dev0+git20240206.afbb9e1
C compiler for the host machine: arm64-apple-darwin20.0.0-clang (clang 16.0.6 "clang version 16.0.6")
C linker for the host machine: arm64-apple-darwin20.0.0-clang ld64 609
C++ compiler for the host machine: arm64-apple-darwin20.0.0-clang++ (clang 16.0.6 "clang version 16.0.6")
C++ linker for the host machine: arm64-apple-darwin20.0.0-clang++ ld64 609
Cython compiler for the host machine: cython (cython 3.0.8)
Host machine cpu family: aarch64
Host machine cpu: aarch64
Program python found: YES (/private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/build-env-9ygl4v_7/bin/python)
Found pkg-config: /opt/homebrew/bin/pkg-config (0.29.2)
Run-time dependency python found: YES 3.11
Has header "Python.h" with dependency python-3.11: YES 
Compiler for C supports arguments -fno-strict-aliasing: YES (cached)
Compiler for C supports arguments -ftrapping-math: YES (cached)
Compiler for C supports link arguments -Wl,-ld_classic: NO (cached)
...
```

</summary>

**Comment by mattip** [comment]
The windows builds are failing because PKG_CONFIG_PATH is not set properly (and because of https://github.com/MacPython/openblas-libs/issues/137). I cannot figure out an incantation that will expand `$PWD/.openblas` in the environment section of the pyproject.toml. I may need to resort to setting the value from a `CIBW_ENVIRONMENT_WINDOWS` environment variable in the github action.

**Comment by mattip** [comment]
bleh. Setting the `PKG_CONFIG_PATH=${{ github.workspace }}/.openblas` on windows  [results in](https://github.com/numpy/numpy/actions/runs/7806986526/job/21294448868?pr=25505#step:6:269) `PKG_CONFIG_PATH D:anumpynumpy/.openblas`. I need to find a way to replace '\' with '/'.

There is a [failing test](https://cirrus-ci.com/task/5603132542025728?logs=build#L3197) on macos arm64 monterey cp310, but only in that build. Is it flaky?
```
FAILED _core/tests/test_multiarray.py::test_sort_float[f-N66]
```

**Comment by andyfaff** [comment]
I think the test failure is sporadic, but it's the same one each time it happens (I think). A test shouldn't cause the test thread to crash though!

**Comment by charris** [comment]
>Is it flaky?

I've seen it before, but not consistently.

**Comment by andyfaff** [comment]
When working in those scripts you have to use this sort of path spec, https://github.com/scipy/scipy/blob/5892d5e1bbdb94d27e0e33c9cff3d0060204ab43/tools/wheels/cibw_before_build_win.sh#L13

File: tools/wheels/LICENSE_win32.txt [hunk 5]
```diff
@@ -96,7 +96,7 @@ License: BSD-3-Clause-Attribution
 
 
 Name: GCC runtime library
-Files: numpy.libs\libopenblas*.dll
+Files: numpy.libs\libgfortran*.dll
```
Comment (mattip) [inline tools/wheels/LICENSE_win32.txt:99]: This was a typo

File: tools/wheels/cibw_before_build.sh [hunk 6]
```diff
@@ -1,7 +1,10 @@
 set -xe
 
-PROJECT_DIR="$1"
-PLATFORM=$(PYTHONPATH=tools python -c "import openblas_support; print(openblas_support.get_plat())")
+PROJECT_DIR="${1:-$PWD}"
+
+
+# remove any cruft from a previous run
+git clean -C $PROJECT_DIR -xfdd
```
Comment (mattip) [inline tools/wheels/cibw_before_build.sh:7]: needed for building more than one wheel in one cibuildwheel run

File: pyproject.toml [hunk 7]
```diff
@@ -155,23 +154,37 @@ manylinux-x86_64-image = "manylinux2014"
...
+[tool.cibuildwheel.macos.environment]
+# In cibuildwheel, {project} is $PWD
+RUNNER_OS="macOS"
+PKG_CONFIG_PATH="$PWD/.openblas"
+DYLD_LIBRARY_PATH="$DYLD_LIBRARY_PATH:/$PWD/.openblas/lib"
 
 [tool.cibuildwheel.windows]
-environment = {PKG_CONFIG_PATH="C:/opt/64/lib/pkgconfig"}
-config-settings = "setup-args=--vsenv setup-args=-Duse-ilp64=true setup-args=-Dblas=openblas setup-args=-Dlapack=openblas"
+# This does not work, use CIBW_ENVIRONMENT_WINDOWS
```
Comment (andyfaff) [inline pyproject.toml:180]: It's not clear to me what this comment means.

Comment (mattip) [reply pyproject.toml:180]: using `./.openblas` for `PKG_CONFIG_PATH` is not good enough. It needs to be a full path since (apparently) the cibuildwheel build does a `cd` before the build. On other OSes, I can use `$PWD`, but not on windows. So I needed to capture the workspace directory in an environment variable in the `windows.yml` github workflow.

File: tools/wheels/cibw_before_build.sh [hunk 8]
```diff
@@ -16,54 +19,36 @@ elif [[ $RUNNER_OS == "Windows" ]]; then
...
-
-if [[ $RUNNER_OS == "macOS" && ("$INSTALL_GFORTRAN" = true) ]]; then
-    # Install same version of gfortran as the openblas-libs builds
-    if [[ $PLATFORM == "macosx-arm64" ]]; then
-        PLAT="arm64"
-    fi
-
-    source $PROJECT_DIR/tools/wheels/gfortran_utils.sh
-    install_gfortran
-    pip install "delocate==0.10.4"
```
[the remark this answers is not in the document]
Comment (rgommers) [reply tools/wheels/cibw_before_build.sh:68]: > but also that we cannot test 2py in these builds.

That's fine; those are way too slow and need their own CI jobs instead. There's already an open issue for that somewhere.

**Comment by mattip** [comment]
I think this is ready for review. Some follow up, that I will do together with fixes from review:
- delete the `tools/openblas_support.py` script
- think about compacting the build matrix for [cirrus arm64 builds](https://cirrus-ci.com/build/5355164987555840).  Even the double `cp310 cp311` build is finishing in ~7 minutes.

**Comment by mattip** [comment]
One wheel build [timed out in testing](https://github.com/numpy/numpy/actions/runs/7825943909/job/21350955158#step:6:3015). Everything else passed. I don't think the failure is connected to this PR, but maybe?

**Comment by mattip** [comment]
Rerunning the timed-out muslinux wheel build, it now passed. CI is green.

**Comment by andyfaff** [comment]
A question before I review. The TLDR of the PR seems to be: install scipy_openblasX, then create a PKG_CONFIG_PATH with a scipy-openblas.pc file that is sourced from the scipy-openblasX wheel.

It's probably been discussed before, but it's not clear to me why the build (i.e. meson) can't be scripted in such a way that if it knows it's going to be using the scipy-openblasX wheel to source openblas, then why can't it query the wheel to figure out where the pc file is, and where the shared library is? i.e. why does this PR have to jump through hoops to move the pc file into a PKG_CONFIG_PATH?

e.g. on a mac the spin cmd for build appears to work as follows (won't have all the details correct here):

1. If there are no blas specified find and use Accelerate
2. If you use the `-use-scipy-openblas` flag then use openblas from the wheel
3. use pkg-config to figure out the relevant BLAS, possibly having to use flag settings to change which BLAS is desired, and PKG_CONFIG_PATH to specify where the pc file is.

My point is that in step 2 the user doesn't have to fiddle about with setting PKG_CONFIG_PATH, or creating files, everything is taken care of behind the scenes. Why can't this happen  in this PR, why the manual intervention by setting paths beforehand?

File: .github/workflows/linux.yml [hunk 9]
```diff
@@ -186,15 +186,21 @@ jobs:
...
-        sudo cp -r $target/lib/* /usr/lib
-        sudo cp $target/include/* /usr/include
+        mkdir -p $PKG_CONFIG_PATH
+        python -m pip install scipy-openblas64
+        pkg_config=$(python -c "import scipy_openblas64; print(scipy_openblas64.get_pkg_config())")
+        echo writing to $PKG_CONFIG_PATH/scipy-openblas.pc
+        echo $pkg_config > $PKG_CONFIG_PATH/scipy-openblas.pc
     - name: Build a wheel via an sdist
+      env:
+        PKG_CONFIG_PATH: ${{ github.workspace }}/.openblas
```
Comment (andyfaff) [inline .github/workflows/linux.yml:203]: This is set twice, probably because environment variables aren't persistent between steps? Would it be better to use [GITHUB_ENV](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable)?

File: .github/workflows/linux_blas.yml [hunk 10]
```diff
@@ -72,26 +72,32 @@ jobs:
         python-version: '3.11'
 
     - name: Install dependencies
+      env:
+        # Need to do this in every step since setup-python overrides PKG_CONFIG_PATH
+        PKG_CONFIG_PATH: ${{ github.workspace }}/.openblas
```
Comment (andyfaff) [inline .github/workflows/linux_blas.yml:77]: Write to GITHUB_ENV to prevent having to set it repeatedly?

File: .github/workflows/linux.yml (hunk 9 above)
Comment (andyfaff) [reply .github/workflows/linux.yml:203]: Like you set [here](https://github.com/numpy/numpy/pull/25505/files#diff-75627befeafda60a526c116b32f6fdef9bf57bcf456f840d7592c1d6f13facb4R118)

File: pyproject.toml [hunk 11]
```diff
@@ -155,23 +154,37 @@ manylinux-x86_64-image = "manylinux2014"
 manylinux-aarch64-image = "manylinux2014"
 musllinux-x86_64-image = "musllinux_1_1"
 # RUNNER_OS is a GitHub Actions specific env var; define it here so it works on Cirrus CI too
-environment = {RUNNER_OS="Linux"}
+[tool.cibuildwheel.linux.environment]
+RUNNER_OS="Linux"
+PKG_CONFIG_PATH="/project/.openblas"
```
Comment (andyfaff) [inline pyproject.toml:159]: This appears to be GHA specific. Can the pc file be placed in a known location in CIBW_BEFORE_BUILD? What happens with the cirrus linux_aarch64 build which has a different directory structure to GHA?

File: tools/wheels/cibw_before_build.sh [hunk 12]
```diff
@@ -16,54 +19,36 @@ elif [[ $RUNNER_OS == "Windows" ]]; then
...
-        cp $basedir/include/* /usr/local/include
-    fi
-elif [[ $RUNNER_OS == "Windows" ]]; then
+# Install Openblas from scipy-openblas64
+if [[ "$INSTALL_OPENBLAS" = "true" ]] ; then
+    echo PKG_CONFIG_PATH $PKG_CONFIG_PATH
+    PKG_CONFIG_PATH=$PROJECT_DIR/.openblas
+    rm -rf $PKG_CONFIG_PATH
+    mkdir -p $PKG_CONFIG_PATH
+    python -m pip install scipy-openblas64
```
Comment (andyfaff) [inline tools/wheels/cibw_before_build.sh:36]: for maintenance branches how do we choose what version of scipy-openblas gets installed. e.g. for a certain maintenance branch we may want to stick with a certain (older) version, and bump when we do a major release. At the moment we just get the latest version.

File: tools/wheels/cibw_before_build.sh [hunk 13]
```diff
@@ -16,54 +19,36 @@ elif [[ $RUNNER_OS == "Windows" ]]; then
...
-        sudo cp -r $basedir/lib/* /usr/local/lib
-        sudo cp $basedir/include/* /usr/local/include
-    else
-        cp -r $basedir/lib/* /usr/local/lib
-        cp $basedir/include/* /usr/local/include
-    fi
-elif [[ $RUNNER_OS == "Windows" ]]; then
+# Install Openblas from scipy-openblas64
+if [[ "$INSTALL_OPENBLAS" = "true" ]] ; then
+    echo PKG_CONFIG_PATH $PKG_CONFIG_PATH
```
Comment (andyfaff) [inline tools/wheels/cibw_before_build.sh:32]: So we have to install scipy-openblas here - understood. However, why is the rest needed? Can't the meson build process do the querying of scipy-openblas to figure out all the necessary info?

File: .github/workflows/linux.yml (hunk 9 above)
Comment (mattip) [reply .github/workflows/linux.yml:203]: Good point.

File: tools/wheels/cibw_before_build.sh (hunk 12 above)
Comment (mattip) [reply tools/wheels/cibw_before_build.sh:36]: This should use the ci_requirements.txt file.

File: tools/wheels/cibw_before_build.sh (hunk 13 above)
Comment (mattip) [reply tools/wheels/cibw_before_build.sh:32]: Meson could run python and try to import scipy_openblas32, then try to import scipy_openblas64. If either of those succeed, it could use `get_include()` and `get_libary()` to add the proper flags to the build. @rgommers thoughts?

File: pyproject.toml (hunk 11 above)
Comment (mattip) [reply pyproject.toml:159]: `/project` is a cibuildwheel convention, not a GHA one. It is a known path inside the docker image which is always used to build on linux with cibuildwheel.

**Comment by mattip** [comment]
> Why can't this happen in this PR, why the manual intervention

The cibw_before_build.sh script does exactly what `spin --with-open-blas` does: create a `pc` file. Using spin during the build would not be a 1:1 replacement. In wheel building, the wheel packaging step (delocate/auditwheel/delvewheel) needs to know where to find the shared objects, which must be done outside the controlled `spin` environment. I guess this too could be moved inside `spin` with a dedicated `spin repair-wheel` command to abstract out the differences between the three tools and to add the proper path to the shared objects.

Is that what you were thinking?

File: pyproject.toml (hunk 11 above)
Comment (andyfaff) [reply pyproject.toml:159]: Worth a comment?

**Comment by mattip** [comment]
> the spin cmd for build appears to work as follows ...

Just to be more accurate: that is not the spin logic, it is logic inside `numpy/meson.build`, which has branching logic to handle the various BLAS options. It could call incantations of `run_command(py, ...)` to replace parts of the `cibw_before_build.sh` script and the `spin --with-scipy-openblas` logic with calls to `scipy_openbas32` or `scipy_openblas64`. I am not sure where that "magic" should happen. 

Outside meson.build: meson tries very hard to not incorporate too much logic and magic inside it, and depends on the user to properly set up pkg-config or cmake files for dependency detection. 

Inside meson.build: there is some merit to not repeating the pkg-config file creation twice, and to eliminate the need to externally set a `PKGC_CONFIG_PATH` environment variable.

**Comment by andyfaff** [comment]
OK, so you're saying that the main purpose is to let delocate/auditwheel/delvewheel know where the libraries are? I don't know why they would need to be told that. For example, if I use `otool` on a recently built scipy dylib:
```
(dev3) 192-168-1-107:linalg andrew$ otool -L cython_lapack.cpython-310-darwin.so 
cython_lapack.cpython-310-darwin.so:
	/opt/arm64-builds/lib/libopenblas.0.dylib (compatibility version 0.0.0, current version 0.0.0)
```
It tells me where the openblas library is! They should just go and grab it.

Perhaps I'm being too nitpicky. What we have so far is obviously working, so perhaps we should merge instead of making things too magical.

**Comment by mattip** [comment]
> I don't know why they would need to be told that.

The tarball used by `openblas_support` is built with absolute paths burned into the shared objects (RPATH and friends) on linux and macos. Using relative paths instead like the wheel build has implications. You can see the commit history where I tried many iterations of things until I got to this point where everything JustWorks and CI is green.

**Comment by andyfaff** [comment]
Ah ok. Well then since it's green we may as well merge, modulo the things that I suggested that you think should be added.

**Comment by mattip** [comment]
I pushed a fix commit:
- remove the tools/openblas_support.py script
- reviewed all the places PKG_CONFIG_PATH is set and try to compact it to once per CI file
- used `spin config-openblas` where possible
- add a comment about `/project` as a path
- compact the cirrus wheel build jobs from 3 to 2. I think it could even be 1 job and still not exceed an hour.

**Comment by mattip** [comment]
Using `spin config-openblas` cannot be done in a wheel-based build, since it prepares the build for the case of using the scipy-openblas __wheel__ at runtime, and in the wheel-build case we want to build a stand-alone wheel that packages the __shared objects__ into the wheel rather than get them from the scipy-openblas wheel.

**Comment by mattip** [comment]
Ci is green after the changes.

**Comment by andyfaff** [comment]
Looks good to me. A good block of work, thanks @mattip

File: pyproject.toml [hunk 14]
```diff
@@ -143,35 +143,50 @@ tracker = "https://github.com/numpy/numpy/issues"
 # `tools/ci/cirrus_wheels.yml`.
 build-frontend = "build"
 skip = "cp36-* cp37-* cp-38* pp37-* *-manylinux_i686 *_ppc64le *_s390x"
-build-verbosity = "3"
 before-build = "bash {project}/tools/wheels/cibw_before_build.sh {project}"
 # The build will use openblas64 everywhere, except on arm64 macOS >=14.0 (uses Accelerate)
-config-settings = "setup-args=-Duse-ilp64=true setup-args=-Dallow-noblas=false build-dir=build"
```
Comment (rgommers) [inline pyproject.toml:149]: @mattip why did you remove `use-ilp64` here? It looks like it switched all the wheels to LP64. Log extracts:
```
# manylinux x86-64:
  Run-time dependency scipy-openblas found: YES 0.3.26
  Message: BLAS symbol suffix: 
```

```
#  Accelerate:
Run-time dependency accelerate found: YES
Message: BLAS symbol suffix: $NEWLAPACK
```

**Comment by rgommers** [comment]
No more tarballs or `openblas_support` script, nice! Thanks for getting this in. Looks good overall, modulo my one comment.

File: pyproject.toml (hunk 14 above)
Comment (mattip) [reply pyproject.toml:149]: I moved the macros to the pkg-config file [here](https://github.com/MacPython/openblas-libs/blob/d65af288e71e4a2099120d18a9cfbaabd77b008d/local/scipy_openblas64/__init__.py#L73) so it is not needed for the scipy-openblas build. I guess it is important to leave it in for diagnostics and for other blas builds though, right?

Comment (rgommers) [reply pyproject.toml:149]: yes indeed
